File: F17DataPaging/F17DataPaging/ViewController.m
//
//  ViewController.m
//  F17DataPaging
//
//  ViewController.m
//  F16RetrievingData
/*
 * *********************************************************************************************************************
 *
 *  BACKENDLESS.COM CONFIDENTIAL
 *
 *  ********************************************************************************************************************
 *
 *  Copyright 2015 BACKENDLESS.COM. All Rights Reserved.
 *
 *  NOTICE: All information contained herein is, and remains the property of Backendless.com and its suppliers,
 *  if any. The intellectual and technical concepts contained herein are proprietary to Backendless.com and its
 *  suppliers and may be covered by U.S. and Foreign Patents, patents in process, and are protected by trade secret
 *  or copyright law. Dissemination of this information or reproduction of this material is strictly forbidden
 *  unless prior written permission is obtained from Backendless.com.
 *
 *  ********************************************************************************************************************
 */

#import "ViewController.h"
#import "Backendless.h"
#import "Location.h"
#import "Restaurant.h"

#define BKNDLSS12192 0

// BasicTest
static NSString *APP_ID = @"25C2E37D-8EB6-8CDA-FF37-D6F05AE7EE00";
static NSString *SECRET_KEY = @"675002F9-A2F8-567F-FFB2-3E348B887900";
static NSString *VERSION_NUM = @"v1";
static NSString *HOST_URL = @"http://api.backendless.com";

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    backendless.hostURL = HOST_URL;
    [backendless initApp:APP_ID secret:SECRET_KEY version:VERSION_NUM];
    
#if BKNDLSS12192
    // http://support.backendless.com/topic/custom-class-name
    [[backendless data] mapTableToClass:@"Restaurants" type:[Restaurant class]];
#endif
    
    [self addRestaurants];

    [self basicPaging];
    [self basicPagingAsync];
    
    [self advancedPaging];
    [self advancedPagingAsync];
}

#pragma mark -
#pragma mark Private Methods

#define PAGESIZE 100

-(void)addRestaurants {
    
    @try {
        
        id <IDataStore> dataStore = [backendless.persistenceService of:[Restaurant class]];
        
        for (int i = 0; i < 300; i++) {
            Restaurant *r = [Restaurant new];
            r.name = [NSString stringWithFormat:@"TastyBaaS %d", i];
            r.cuisine = @"mBaaS";
            r = [dataStore save:r];
            NSLog(@"Save restaurant name = %@", r.name);
        }
    }
    
    @catch (Fault *fault) {
        NSLog(@"Server reported an error: %@", fault);
    }
}

-(void)basicPaging {
    
    @try {
        
        NSDate *startTime = [NSDate date];
        
        BackendlessDataQuery *query = [BackendlessDataQuery query];
        BackendlessCollection *restaurants = [[backendless.persistenceService of:[Restaurant class]] find:query];
        NSLog(@"Total restaurants in the Backendless storage - %@ <'%@'=='%@'>", [restaurants getTotalObjects], [restaurants type], [restaurants getEntityName]);
        
        unsigned long size = 0;
        while ( (size = [[restaurants getCurrentPage] count]) ) {
            NSLog(@"Loaded %lu restaurant in the current page", size);
            restaurants = [restaurants nextPage];
        }
        
        NSLog(@"Total time (ms) - %g", 1000*[[NSDate date] timeIntervalSinceDate:startTime]);
    }
    
    @catch (Fault *fault) {
        NSLog(@"Server reported an error: %@", fault);
    }
}

-(void)nextPageAsync:(BackendlessCollection *)restaurants start:(NSDate *)start {
    
    unsigned long size = [[restaurants getCurrentPage] count];
    if (!size) {
        NSLog(@"Total time (ms) - %g", 1000*[[NSDate date] timeIntervalSinceDate:start]);
        return;
    }
    
    NSLog(@"Loaded %lu restaurant in the current page", size);
    
    [restaurants
     nextPageAsync:^(BackendlessCollection *rests) {
         [self nextPageAsync:rests start:start];
     }
     error:^(Fault *fault) {
         NSLog(@"Server reported an error: %@", fault);
     }];
}

-(void)basicPagingAsync {
    
    NSDate *startTime = [NSDate date];
    
    BackendlessDataQuery *query = [BackendlessDataQuery query];
    [[backendless.persistenceService of:[Restaurant class]]
     find:query
     response:^(BackendlessCollection *restaurants) {
         NSLog(@"Total restaurants in the Backendless storage - %@", [restaurants getTotalObjects]);
         [self nextPageAsync:restaurants start:startTime];
     }
     error:^(Fault *fault) {
         NSLog(@"Server reported an error: %@", fault);
     }];
}

-(void)advancedPaging {
    
    @try {
        
        NSDate *startTime = [NSDate date];
        
        BackendlessDataQuery *query = [BackendlessDataQuery query];
        query.queryOptions.relationsDepth = @(1);
        query.queryOptions.pageSize = @(PAGESIZE); //set page size
        BackendlessCollection *restaurants = [[backendless.persistenceService of:[Restaurant class]] find:query];
        NSLog(@"Total restaurants in the Backendless storage - %@", [restaurants getTotalObjects]);
        
        int offset = 0;
        unsigned long size = 0;
        while ( (size = [[restaurants getCurrentPage] count]) ) {
            NSLog(@"Loaded %lu restaurant in the current page", size);
            offset += size;
            restaurants = [restaurants getPage:offset pageSize:PAGESIZE];
        }
        
        NSLog(@"Total time (ms) - %g", 1000*[[NSDate date] timeIntervalSinceDate:startTime]);
    }
    
    @catch (Fault *fault) {
        NSLog(@"Server reported an error: %@", fault);
    }
}

-(void)getPageAsync:(BackendlessCollection *)restaurants offset:(int)offset start:(NSDate *)start {
    
    unsigned long size = [[restaurants getCurrentPage] count];
    if (!size) {
        NSLog(@"Total time (ms) - %g", 1000*[[NSDate date] timeIntervalSinceDate:start]);
        return;
    }
    
    NSLog(@"Loaded %lu restaurant in the current page", size);
    
    offset += size;
    [restaurants
     getPage:offset pageSize:PAGESIZE
     response:^(BackendlessCollection *rests) {
         [self getPageAsync:rests offset:offset start:start];
     }
     error:^(Fault *fault) {
         NSLog(@"Server reported an error: %@", fault);
     }];
}

-(void)advancedPagingAsync {
    
    NSDate *startTime = [NSDate date];
    
    int offset = 0;
    BackendlessDataQuery *query = [BackendlessDataQuery query];
    query.queryOptions.relationsDepth = @(1);
    query.queryOptions.pageSize = @(PAGESIZE); //set page size
    [[backendless.persistenceService of:[Restaurant class]]
     find:query
     response:^(BackendlessCollection *restaurants) {
         
         NSLog(@"Total restaurants in the Backendless storage - %@", [restaurants getTotalObjects]);
         [self getPageAsync:restaurants offset:offset start:startTime];
     }
     error:^(Fault *fault) {
         NSLog(@"Server reported an error: %@", fault);
     }];
}

@end

File: F17DataPaging/F17DataPagingSwift/ViewController.swift
//
//  ViewController.swift
//  F17DataPagingSwift
//
//  ViewController.swift
//  F16RetrievingDataSwift
/*
* *********************************************************************************************************************
*
*  BACKENDLESS.COM CONFIDENTIAL
*
*  ********************************************************************************************************************
*
*  Copyright 2015 BACKENDLESS.COM. All Rights Reserved.
*
*  NOTICE: All information contained herein is, and remains the property of Backendless.com and its suppliers,
*  if any. The intellectual and technical concepts contained herein are proprietary to Backendless.com and its
*  suppliers and may be covered by U.S. and Foreign Patents, patents in process, and are protected by trade secret
*  or copyright law. Dissemination of this information or reproduction of this material is strictly forbidden
*  unless prior written permission is obtained from Backendless.com.
*
*  ********************************************************************************************************************
*/

import UIKit

class ViewController: UIViewController {
    
    // BasicTest
    let APP_ID = "25C2E37D-8EB6-8CDA-FF37-D6F05AE7EE00"
    let SECRET_KEY = "675002F9-A2F8-567F-FFB2-3E348B887900"
    let VERSION_NUM = "v1"
    
    let PAGESIZE = 100
    
    var backendless = Backendless.sharedInstance()
    var offset = 0
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        backendless.initApp(APP_ID, secret:SECRET_KEY, version:VERSION_NUM)
        
        //addRestaurants()
        
        //basicPaging()
        basicPagingAsync()
        
        //advancedPaging()
        //advancedPagingAsync()
    }
    
    func addRestaurants() {
        
        Types.tryblock({ () -> Void in
            
            let dataStore = self.backendless.persistenceService.of(Restaurant.ofClass())
            for i in 0 ..< 300 {
                var r = Restaurant()
                r.name = "TastyBaaS \(i)"
                r.cuisine = "mBaaS"
                r = dataStore.save(r) as! Restaurant
                print("Save restaurant name = \(r.name)")
            }
            },
            
            catchblock: { (exception) -> Void in
                print("Server reported an error: \(exception as! Fault)")
            }
        )
    }
    
    func basicPaging() {
        
        Types.tryblock({ () -> Void in
            
            let startTime = NSDate()
            
            let query = BackendlessDataQuery()
            var restaurants = self.backendless.persistenceService.of(Restaurant.ofClass()).find(query)
            print("Total restaurants in the Backendless starage - \(restaurants.totalObjects)")
            
            var size = restaurants.getCurrentPage().count
            while size > 0 {
                print("Loaded \(size) restaurant in the current page")
                restaurants = restaurants.nextPage()
                size = restaurants.getCurrentPage().count
            }
            
            print("Total time (ms) - \(1000*NSDate().timeIntervalSinceDate(startTime))")
            },
            
            catchblock: { (exception) -> Void in
                print("Server reported an error: \(exception as! Fault)")
            }
        )
    }
    
    func nextPageAsync(restaurants: BackendlessCollection, startTime: NSDate) {
        
        let size = restaurants.getCurrentPage().count
        if size == 0 {
            print("Total time (ms) - \(1000*NSDate().timeIntervalSinceDate(startTime))")
            return
        }
        
        print("Loaded \(size) restaurant in the current page")
        let currentPage = restaurants.getCurrentPage() as! [Restaurant]
        for restaurant in currentPage {
            if restaurant.name != nil {
                print(restaurant.name)
            }
        }
        
        restaurants.nextPageAsync(
            { ( rests : BackendlessCollection!) -> () in
                self.nextPageAsync(rests, startTime:startTime)
            },
            error: { ( fault : Fault!) -> () in
                print("Server reported an error: \(fault)")
            }
        )
    }
    
    func basicPagingAsync() {
        
        let startTime = NSDate()
        
        let query = BackendlessDataQuery()
        backendless.persistenceService.of(Restaurant.ofClass()).find(
            query,
            response: { ( restaurants : BackendlessCollection!) -> () in
                print("Total restaurants in the Backendless starage - \(restaurants.totalObjects)")
                self.nextPageAsync(restaurants, startTime:startTime)
            },
            error: { ( fault : Fault!) -> () in
                print("Server reported an error: \(fault)")
            }
        )
    }
    
    func advancedPaging() {
        
        Types.tryblock({ () -> Void in
            
            let startTime = NSDate()
            
            let query = BackendlessDataQuery()
            query.queryOptions.pageSize = self.PAGESIZE // set page size
            var restaurants = self.backendless.persistenceService.of(Restaurant.ofClass()).find(query)
            print("Total restaurants in the Backendless starage - \(restaurants.totalObjects)")
            
            var offset = 0
            var size = restaurants.getCurrentPage().count
            while size > 0 {
                print("Loaded \(size) restaurant in the current page")
                offset += size
                restaurants = restaurants.getPage(offset, pageSize:self.PAGESIZE)
                size = restaurants.getCurrentPage().count
            }
            
            print("Total time (ms) - \(1000*NSDate().timeIntervalSinceDate(startTime))")
            },
            
            catchblock: { (exception) -> Void in
                print("Server reported an error: \(exception as! Fault)")
            }
        )
    }
    
    func getPageAsync(restaurants: BackendlessCollection, startTime: NSDate) {
        
        let size = restaurants.getCurrentPage().count
        if size == 0 {
            print("Total time (ms) - \(1000*NSDate().timeIntervalSinceDate(startTime))")
            return
        }
        
        print("Loaded \(size) restaurant in the current page")
        
        self.offset += size
        restaurants.getPage(
            self.offset,
            pageSize:PAGESIZE,
            response: { ( rests : BackendlessCollection!) -> () in
                self.getPageAsync(rests, startTime:startTime)
            },
            error: { ( fault : Fault!) -> () in
                print("Server reported an error: \(fault)")
            }
        )
    }
    
    func advancedPagingAsync() {
        
        let startTime = NSDate()
        
        self.offset = 0
        let query = BackendlessDataQuery()
        query.queryOptions.pageSize = PAGESIZE // set page size
        backendless.persistenceService.of(Restaurant.ofClass()).find(
            query,
            response: { ( restaurants : BackendlessCollection!) -> () in
                print("Total restaurants in the Backendless starage - \(restaurants.totalObjects)")
                self.getPageAsync(restaurants, startTime:startTime)
            },
            error: { ( fault : Fault!) -> () in
                print("Server reported an error: \(fault)")
            }
        )
    }
    
}

